### ISSUE TITLE
Upgrading internals to use terra

### ISSUE THREAD

**Issue by Nowosad** [r-spatialecology/landscapemetrics#245]
I think we should start to consider updating the landscapemetrics internals to use terra instead of raster (raster uses rgdal), and sf instead of sp - please read https://www.mail-archive.com/r-sig-geo@r-project.org/msg18468.html.

@mhesselbarth @bitbacchus what do you think?

**Comment by bitbacchus** [comment]
I agree. We already have terra support, don't we? And internally we use plain matrices, so there is probably not too much to do.

Also, we can just support raster, as long as there is demand (aka "suggests" or "enhances").

**Comment by mhesselbarth** [comment]
I don't really understand this. All calculates are done using a matrix anyhow. So you want to return a terra object whenever we return a raster? Or what do you understand as internals?

**Comment by Nowosad** [comment]
I would need to look deeper in the code, but now I am thinking about:

1. Returning terra SpatRaster instead of raster objects
2. Not using raster functions - I know that we, at least, use one - `raster::ncell()`
3. Similarly, not using sp functions - we are doing that, for example, in `sample_lsm()`
4. Updates imports and suggests in DESCRIPTION based on these changes
5. (Optional) consider using fast terra code in some places

Also, we do not need to do it now - some time next year should be good.

**Comment by mhesselbarth** [comment]
Sounds all good to me. I think shouldn't be too much work because as said, for most functions we convert to a matrix first and go on from there

**Comment by bitbacchus** [comment]
I worked a bit on the terra branch with the idea to move from `raster::focal` to `terra::focal` because `terra::focal` supports vectors as output for the focal function. I think this is useful for our aim to support class-level metrics. I ran into some troubles with internal raster::functions using `terra::SpatRasters` (not always sure why). My workaround in such cases is basically:
```r
if (inherits(x = x, what = "RasterLayer")) {
  # raster::fun
} else {
  # terra::fun
}
```
My aim was that the functions return a `SpatRaster` when they get a `SpatRaster` as input.

Something I stumbled over and don't understand is this line: 
https://github.com/r-spatialecology/landscapemetrics/blob/acd3fd6dffa1908a8afb3cfaa5341cc50054e1ae/R/show_cores.R#L100 

Why does `raster::crop` get directions as an argument? Does it do anything?

**Comment by mhesselbarth** [comment]
No idea, but I want to say that doesn't need to be there? I think its not getting used anyhow since the `...` arguments of `crop` are forwarded to `writeRaster`? This all doesn't make sense 🤷‍♂️

**Comment by mhesselbarth** [comment]
@Nowosad @bitbacchus 

I worked a bit on the `terra` branch today. However, I ran into a conceptual problem quite fast. 

The issue is, that `SpatRaster` cannot be saved as `.rda` files, but must be `terra::pack(x)` before saving. When loading, the object must be wrapped into `terra::rast()` before being used. This is due to the internal data behavior of `terra`. However, this is  an issue for the example data we are including. Basically, we would need to exchange all calls to `landscapemetrics::landscape` into `terra::rast(landscapemetrics::landscape)`. 

I tried to also play around with an `.onload()` function, to unpack the file each time the package is loaded. However, the issue with that is, I am struggling the "replace" the packed raster. Thus we would be able to unpack `landscape_pack` into `landscape` for example, however, the original included data would obviously still be packed...

**Comment by mhesselbarth** [comment]
Alternatively we could use e.g., `.tif` files in inst/extdata, but similar this would then need some pre-processing (i.e., reading the `tif`-file) before the raster would be useable

**Comment by Nowosad** [comment]
Hi @mhesselbarth, I can think of three possible approaches for example data reading:

```r
# remotes::install_github("nowosad/test")

# 1
# Depends: terra
library(test)
test:::landscape
plot(test:::landscape)

landscape = test:::landscape

# 2
# Imports: terra
library(test)
library(terra)
test:::landscape
plot(test:::landscape)

landscape = test:::landscape

# 3
library(test)
library(terra)
landscape = rast(system.file("raster/landscape.tif", package = "test"))
plot(landscape)
```

In general, I use the last one in some other of my packages (as well in the Geocomputation with R e2 book), but I am open to using any alternative. 

I also do not think this is a major issue. More importantly, I think we should figure out a depreciation approach. Currently, the package accepts a lot of input classes, but returns Raster objects for some functions. It also accepts sp objects in some cases. I think we should have at least two landscapemetrics releases with messages that things will change. For example, in the first release, there could be a starting message that the landscapemetrics will switch to terra in mid-2023. Then, in the second release, there should be some messages/warnings in some functions, e.g., `spatialize_lsm()`, that their behavior will be different soon. We could release the terra version in mid-2023, with messages about the change in some places. Then, we could remove these messages a few months later... Let me know what you think about it.

**Comment by mhesselbarth** [comment]
Considering the example data, I think the third approach is the best. This is also what people in the ROpenSci Slack suggested. But I agree, that is not the major discussion point. 

In terms of input, my idea was to "flip" the defaults. So take a `terra` object as the default, but still allow `raster` and convert them. Basically the same we do now but vice versa. 

Actually, I think there won't be too many major breaking changes cause the majority of the output are `tibbles` anyhow. As you said, it's mostly `spatialize_lsm()` and related functions that return `raster` at the moment. I like the idea of having a message/warning for a bit. 

However, I think two releases might be even too much. Especially, since users can convert back to `raster` with one line of code. I would suggest having a warning in all functions that return a `raster` and also adding the information to all `README`s etc.

**Comment by Nowosad** [comment]
- Input - yes, completely agree -- we just can do that invisibly to the users
- Output - yep, I am fine with the warnings

**Comment by mhesselbarth** [comment]
Sounds all good to me. I probably have some time to start working on this the next days!

**Comment by Nowosad** [comment]
Awesome. If I is possible, please also consider the point from https://github.com/r-spatialecology/landscapemetrics/issues/224#issuecomment-1024424440.

**Comment by mhesselbarth** [comment]
Reminder to myself: 

- [ ] Make sure extract and sample support lines and polygons again (including tests)
- [ ] Make sure sf class checks work properly

**Comment by mhesselbarth** [comment]
Small update: On the terra branch everything should run now more or less with `raster` or `sp`. Also, on `main` and the next CRAN release, there will be a message that `raster` support will be remove soon.

Main thing to still improve is everything that includes `sf` objects, so basically sampling and extracting. That got a bit messy and I want to simplify the code there. Currently only really working for a `matrix`.

**Comment by Nowosad** [comment]
@mhesselbarth 

1. I think the v2 may also be a good moment to follow the discussion at https://github.com/r-spatialecology/landscapemetrics/issues/224#issuecomment-1024424440. 
2. Do you want me to test the terra branch already or should I wait for some new changes?

**Comment by mhesselbarth** [comment]
I think its ready for a first test and playing around with it! 

As I already mentioned above, all extracting and sampling is not really working for anything else but matrices.

**Comment by Nowosad** [comment]
👍🏻 I will try to test in the next few days.

**Comment by Nowosad** [comment]
@mhesselbarth I had a spare hour, and made some pushes to the terra branch. It includes:

1. Improving the starting message (to use the present time).
2. Reinstating Raster* objects support (it is related to only a dozen lines of code + "raster" in "Enhances"). I am convinced that it will make life of many of our users better.

Some things to do:

1. Update the param message from "Raster* Layer, Stack, Brick, SpatRaster (terra), stars, or a list of rasterLayers" to "SpatRaster; Raster* Layer, Stack, Brick; stars; or a list of SpatRasters" (or similar)
2. I think that the use of `terra::unwrap()` in the documentation will confuse many people. Thus, I have a few suggestions (in my order of preferences): 

(a) add a new function, e.g., `read_example_data()`, and then use it before the example code, for example:

```
read_example_data(landscape)
lsm_l_ai(landscape)
```

(b) Just move the `unwrap()` to a new line before the calculations, e.g.,

```
# good
landscape = terra::unwrap(landscape)
lsm_l_ai(landscape)

# bad
lsm_l_ai(terra::unwrap(landscape))
```

(c) Use system.file to read example data. See the last example at https://github.com/r-spatialecology/landscapemetrics/issues/245#issuecomment-1298215147.

I also still plan to spend more time testing in the next few days.

**Comment by mhesselbarth** [comment]
Sounds all good to me. In terms of the unwrap call. I think I either prefer option (b) or (c), but have no strong opinion which one I like more. Maybe slightly leaning towards (b), but dont really care/

**Comment by mhesselbarth** [comment]
@Nowosad 
Did you see any weird behavior yet in terms of `terra`?

**Comment by Nowosad** [comment]
@mhesselbarth there are bugs popping out here and there (https://github.com/rspatial/terra/issues), but in general Robert Hijmans is very response and fixes them quick.

**Comment by mhesselbarth** [comment]
Oh I was referring specifically to any `landscapemetrics` functions...

**Comment by mhesselbarth** [comment]
Please see #285. Given we just update `CRAN` including the start-up message, I think we can merge `terra` in to `main`

**Comment by mhesselbarth** [comment]
The gh-action fails because some packages (`raster`) cant be found anymore. However, I do install them as `extra-packages`. So not sure what's going on. Probably somehow related to the `DESCRIPTION` fields

**Comment by Nowosad** [comment]
I think it was because we had `library(raster)` in a testthat file. It should be fine now, and yes -- we could now merge `terra` to `main`.

**Comment by Nowosad** [comment]
It is ok on my laptop, but seems not to be fine on gh-actions...

**Comment by mhesselbarth** [comment]
Main is now using terra

**Comment by Nowosad** [comment]
pkgdown is down -- https://github.com/r-spatialecology/landscapemetrics/actions/runs/4562632755/jobs/8050099932

**Comment by mhesselbarth** [comment]
Will fix d8893d1a
